python/trainers/ppo_trainer.py

```diff
@@ -0,0 +1,346 @@
+                    if self.use_observations: 
+                        feed_dict[self.model.observation_in] = np.vstack(info.observations)
+                    if self.use_states:
+                        feed_dict[self.model.state_in] = info.states
+                    if self.use_recurrent:
+                        feed_dict[self.model.memory_in] = info.memories
+                    value_next = self.sess.run(self.model.value, feed_dict)[l]
+                agent_id = info.agents[l]
+                self.training_buffer[agent_id]['advantages'].set(
+                    get_gae(
+                        rewards=self.training_buffer[agent_id]['rewards'].get_batch(),
+                        value_estimates=self.training_buffer[agent_id]['value_estimates'].get_batch(),
+                        value_next=value_next, gamma=self.trainer_parameters['gamma'], lambd=self.trainer_parameters['lambd'])
+                )
+                self.training_buffer[agent_id]['discounted_returns'].set( \
+                 self.training_buffer[agent_id]['advantages'].get_batch() \
+                 + self.training_buffer[agent_id]['value_estimates'].get_batch())
+
+                self.training_buffer.append_global(agent_id, batch_size = None, training_length=self.sequence_length)
+
+                self.training_buffer[agent_id].reset_agent()
+                if info.local_done[l]:
+                    self.stats['cumulative_reward'].append(self.cumulative_rewards[agent_id])
+                    self.stats['episode_length'].append(self.episode_steps[agent_id])
+                    self.cumulative_rewards[agent_id] = 0
+                    self.episode_steps[agent_id] = 0
+
+
+    def reset_buffers(self):
+        """
+        A signal that the buffer must be reset. Get typically called when the academy resets.
+        """
+        self.training_buffer.reset_all()
+        for agent_id in self.cumulative_rewards:
+            self.cumulative_rewards[agent_id] = 0
+        for agent_id in self.episode_steps:
+            self.episode_steps[agent_id] = 0
+
+    def is_ready_update(self):
+        """
+        Returns wether or not the trainer has enough elements to run update model
+        :return: A boolean corresponding to wether or not update_model() can be run
+        """
+        return len(self.training_buffer.global_buffer['actions']) > self.trainer_parameters['buffer_size']
+
+    def update_model(self):
+        """
+        Uses training_buffer to update model.
+        :param batch_size: Size of each mini-batch update.
+        :param num_epoch: How many passes through data to update model for.
+        """
+        num_epoch = self.trainer_parameters['num_epoch']
+        batch_size = self.trainer_parameters['batch_size']
+        total_v, total_p = 0, 0
+        advantages = self.training_buffer.global_buffer['advantages'].get_batch()
+        self.training_buffer.global_buffer['advantages'].set(
```

Comment:
it's very brittle for this code to have to know so much about the internals (e.g. keys) of the dictionary used by the buffer class and specifically to have to use string literals copy+pasted throughout the code like 'num_epoch'. Instead, the buffer class should expose a clean API that allows the implementation to change w/o its client (i.e. this code) needing to change.

I don't expect that refactor to be part of this PR, but we should at least track it and add a TODO; otherwise we're just adding technical debt here